public/scripts/views/forms/custom-id.js

```diff
@@ -0,0 +1,164 @@
+                  event.target.setCustomValidity("ID already exists");
+                } else {
+                  event.target.setCustomValidity("");
+                }
+              }, function (e) {
+                event.target.setCustomValidity("");
+              });
+            }
+          } else if (service[0]=='teams' && service[1] == 'getMembership') {
+            var teamId = element.form.elements.namedItem("teamId").value;
+              sdk[service[0]][service[1]](teamId, value).then(function (res) {
+                if (res.$id == value) {
+                  event.target.setCustomValidity("ID already exists");
+                } else {
+                  event.target.setCustomValidity("");
+                }
+              }, function (e) {
+                event.target.setCustomValidity("");
+              });
+          } else {
+            sdk[service[0]][service[1]](value).then(function (res) {
+              if (res.$id == value) {
+                event.target.setCustomValidity("ID already exists");
+              } else {
+                event.target.setCustomValidity("");
+              }
+            }, function (e) {
+              event.target.setCustomValidity("");
+            });
+          }
+        }
+      }
+
+      var setIdType = function (idType) {
+        element.setAttribute("data-id-type", idType);
+        if (idType == "custom") {
+          info.innerHTML = "Allowed Characters A-Z, a-z, 0-9, and non-leading underscore";
+          writer.value = prevData;
+          writer.disabled = false;
+          element.value = prevData;
+          writer.focus();
+          writer.addEventListener('blur', validate);
+        } else {
+          info.innerHTML = "Appwrite will generate a unique ID";
+          prevData = writer.value;
+          writer.disabled = true;
+          writer.value = 'auto-generated';
+          element.value = 'unique()';
+        }
+        button.className = idType == "custom" ? "icon-cog copy" : "icon-edit copy";
+      }
+
+      var sync = function (event) {
+        if (element.value !== 'unique()') {
+          writer.value = element.value;
+        }
+      }
+
+      var syncE = function (event) {
+        element.value = writer.value;
+      }
+
+      var keypress = function (e) {
+        // which key is pressed, keyPressed = e.which || e.keyCode; 
+        const key = e.which || e.keyCode;
+        const ZERO = 48;
+        const NINE = 57;
+        const SMALL_A = 97;
+        const SMALL_Z = 122;
+        const CAPITAL_A = 65;
+        const CAPITAL_Z = 90;
+        const UNDERSCORE = 95;
+
+        const isNotValidDigit = key < ZERO || key > NINE;
+        const isNotValidSmallAlphabet = key < SMALL_A || key > SMALL_Z;
+        const isNotValidCapitalAlphabet = key < CAPITAL_A || key > CAPITAL_Z;
+
+        //Leading underscore is prevented
+        if (key == UNDERSCORE && e.target.value.length == 0) {
+          e.preventDefault();
```

Comment:
Not sure if this is the best UX.

So we don't show any hint at all to why the the user can't type `_`?

Reply:
it's the same for all other disabled keys, we have a helper message below text. It's also same for native maxLength attribute in HTML, it doesn't provide any feedback?

Reply:
Fair enough 👍🏻